worker/credentialvalidator/worker_test.go

```diff
@@ -0,0 +1,163 @@
+	c.Check(err, gc.ErrorMatches, "glug")
+	s.facade.CheckCallNames(c, "ModelCredential", "WatchCredential", "ModelCredential")
+}
+
+func (s *WorkerSuite) TestModelCredentialNotNeeded(c *gc.C) {
+	s.facade.exists = false
+	config := credentialvalidator.Config{
+		Facade: s.facade,
+		Check:  neverCheck,
+	}
+	worker, err := credentialvalidator.New(config)
+	c.Assert(err, gc.ErrorMatches, "model is on the cloud that does not need auth")
+	c.Assert(worker, gc.IsNil)
+	s.facade.CheckCallNames(c, "ModelCredential")
+}
+
+func (s *WorkerSuite) TestCredentialChangeToInvalid(c *gc.C) {
+	s.facade.credentials = []base.StoredCredential{
+		{credentialTag, true},
+		{credentialTag, false},
+	}
+
+	config := credentialvalidator.Config{
+		Facade: s.facade,
+		Check:  credentialvalidator.IsValid,
+	}
+	worker, err := credentialvalidator.New(config)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Check(worker.Check(), jc.IsTrue)
+
+	err = workertest.CheckKilled(c, worker)
+	c.Check(err, gc.Equals, credentialvalidator.ErrChanged)
+	s.facade.CheckCallNames(c, "ModelCredential", "WatchCredential", "ModelCredential")
+}
+
+func (s *WorkerSuite) TestCredentialChangeFromInvalid(c *gc.C) {
+	s.facade.credentials = []base.StoredCredential{
+		{credentialTag, false},
+		{credentialTag, true},
+	}
+
+	config := credentialvalidator.Config{
+		Facade: s.facade,
+		Check:  credentialvalidator.IsValid,
+	}
+	worker, err := credentialvalidator.New(config)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Check(worker.Check(), jc.IsFalse)
+
+	err = workertest.CheckKilled(c, worker)
+	c.Check(err, gc.Equals, credentialvalidator.ErrChanged)
+	s.facade.CheckCallNames(c, "ModelCredential", "WatchCredential", "ModelCredential")
+}
+
+func (s *WorkerSuite) TestModelCredentialReplaced(c *gc.C) {
+	s.facade.credentials = []base.StoredCredential{
+		{credentialTag, true},
+		{names.NewCloudCredentialTag("such/different/credential").String(), false},
+	}
+	config := credentialvalidator.Config{
+		Facade: s.facade,
+		Check:  credentialvalidator.IsValid,
+	}
+	worker, err := credentialvalidator.New(config)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Check(worker.Check(), jc.IsTrue)
+
+	err = workertest.CheckKilled(c, worker)
+	c.Check(err, gc.Equals, credentialvalidator.ErrModelCredentialChanged)
+	s.facade.CheckCallNames(c, "ModelCredential", "WatchCredential", "ModelCredential")
+}
+
+func (s *WorkerSuite) TestNoRelevantCredentialChange(c *gc.C) {
+	s.facade.credentials = []base.StoredCredential{
+		{credentialTag, true},
+		{credentialTag, true},
+		{credentialTag, true},
+		{credentialTag, true},
+	}
+	config := credentialvalidator.Config{
+		Facade: s.facade,
+		Check:  credentialvalidator.IsValid,
+	}
+	worker, err := credentialvalidator.New(config)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Check(worker.Check(), jc.IsTrue)
+
+	workertest.CheckAlive(c, worker)
+	workertest.CleanKill(c, worker)
+	s.facade.CheckCallNames(c, "ModelCredential", "WatchCredential", "ModelCredential", "ModelCredential", "ModelCredential")
```

Comment:
This feels racey to me. How do we know that the right number of calls have been made befor we kill the worker?